### DISCUSSION TITLE
How to do parallel server action calls? My server actions get executed sequentially

### DISCUSSION THREAD

**Discussion by aaronvg** [vercel/next.js#50743] [Help]
### Summary

I need to send like 20 requests, one at a time using a server action. Any ideas on how to do this in parallel? I want to keep track of each request on a client-side component using the useTransition / isPending flag.

**Comment by pingustar** [comment]
I am interested in this as well

**Comment by chungnq253** [comment]
I am interested in this as well!!!

**Comment by VaDominikKienzler** [comment]
Please help

**Comment by Dog-Egg** [comment]
I'd like to know too.

**Comment by dlibis** [comment]
this guy has an answer: https://www.youtube.com/watch?v=CDZg3maL9q0

**Reply by mitchheddles** [reply]
This will work for a small number of requests, but it's not a bulletproof solution.

**Reply by jeroen-ultra** [reply]
This worked for me.

@mitchheddles Can you explain how what happens under the hood when you do this affects the limits for this solution?

**Reply by icflorescu** [reply]
Or this: https://github.com/icflorescu/next-server-actions-parallel 😊

**Comment by jordanpurinton** [comment]
Would love to see this functionality

**Reply by enricoros** [reply]
Agree, could definitely be a flag in `next.config.js`

**Comment by gmchad** [comment]
> this guy has an answer: https://www.youtube.com/watch?v=CDZg3maL9q0

This works locally but in production (hosted on vercel) the requests are still sequential.

**Reply by nicnocquee** [reply]
weird, i tried this but still sequential even in dev local.

**Comment by yonatanlavy** [comment]
Please add this

**Comment by ZoranRavic** [comment]
Creating 20 parallel requests is a bad idea when using serverless functions, since they will likely all encounter cold starts.

Not to mention that it wont work if you have any noticeable number of concurrent users, due to the limit of 1000 active functions at the time.

Your best bet is to rewrite this as a single action.

If you really need to get responses for individual operations as soon as they are done then serverless functions simply aren't a good choice.  
You'll have to setup a normal server or several servers with a load balancer.

**Reply by TanGrumser** [reply]
Give me the chance to do stupid decisions myself.

**Reply by aretrace** [reply]
@TanGrumser 🫡

**Reply by JeroenvdV** [reply]
I'm not sure I agree with the cold starts problem. A sudden burst of calls could all encounter cold starts (as with a single one) if the function is not continuously used, but they also might not if there are regularly a similar amount of concurrent calls. 

I also don't really understand how "serverless functions simply aren't a good choice". I mean, yeah now they aren't for this, as concurrency is not supported, but I don't think there's anything about serverless as an execution model itself that goes against concurrent calls of the same or even different functions. Is this just a chicken and egg thing?

**Reply by ZoranRavic** [reply]
My assumption is that the thing you're primarily doing on the server side is reading from and writing to the database.

In this scenario the bulk of the load is on the database servers, with your server(less) functions mostly just waiting for responses. And if you are on a cheaper database plan, those responses are going to come in slower, since cheap db servers get slower when processing a lot of requests at the exact same time.

You can greatly reduce the load on the database by refactoring your 20 (or however many) requests to be a single api call which would allow you to avoiding duplicate database calls and performing bulk read/write operations, which can be executed more efficiently on the database side than many smaller operations.  
This also greatly reduces your vercel bill, as you are only paying for one function call, instead of 20.

Now that we have a more efficient api we have to find a way to return responses for individual sub-functions as soon as they are done.  
Sadly this can not be done with serverless functions and we have to go back to normal servers.

---

Same logic would apply if you are interacting with third-party apis, for example an api for sending emails or sms messages, since in those cases your function is again just waiting to see if the third-party api returns a 200 OK response.

---

And even in the case you don't care about optimizing your database calls and you are actually interesting in computing on 20 CPU cores you are still better off with a regular server since if you need a lot of compute time then vercel bill is going to put you out of business.

**Comment by DiAifU** [comment]
I'm actually wondering why server actions work like that too ?
I'm currently seeing some of my client-side queries finishing one after the other instead of coming all as fast as possible and I thought server actions would trigger them in parallel and display the results in the browser as soon as it is available.

Why is it implemented like this ? Is it likely to change or should I implement my requests differently ?

**Reply by aretrace** [reply]
> See this last item from the [react docs](https://react.dev/reference/rsc/use-server#caveats), I have yet to see an explanation from the react team as to why they behave this way (I'm sure it's some technical thing due to how react itself it implemented).

![fbimsg](https://github.com/user-attachments/assets/b3cfc807-594e-4c35-8be2-a04ede3ed4d6)

**Reply by ZoranRavic** [reply]
The react docs only say that "frameworks implementing Server Actions typically process one action at a time"

Meaning that this is something the frameworks (notably next.js) decided to do and it's not something that is necessary for react to function properly.

I don't see where did you get your conclusion that react itself couldn't handle parallel action calls.

**Reply by aretrace** [reply]
See this last item from the [react docs](https://react.dev/reference/rsc/use-server#caveats), I have yet to see an explanation from the ~~react~~ vercel team as to why they behave this way (I'm ~~sure~~ guessing it's some technical thing due to how ~~react~~ next.js itself it implemented).

![llms](https://github.com/user-attachments/assets/4744e9a8-984c-4dad-972e-540478279e1e)

**Comment by nikshepsvn** [comment]
This is such a huge gripe when using server-actions, please fix

**Comment by icflorescu** [comment]
Actually, this is quite easy to do with [icflorescu/next-server-actions-parallel](https://github.com/icflorescu/next-server-actions-parallel), if you're looking for a ready-made solution.
Have a look at the repo, the idea simple enough that you can simply copy-paste it in your own projects if you don't want to add it as a dependency.

**Reply by yonatanlavy** [reply]
Is this parallelizing or running sequentially faster, but still waiting for an 'ok' from each one?
From my testing it takes a tiny bit more for this to run on server actions than on routes, so it might still not solve the base problem of actually serializing and on high network latencies will cause huge delays.

Will this work on all environments?

Will this work with existing error handling and payloads including forms, files, components, etc..?

**Reply by icflorescu** [reply]
It's definitely running the actions in parallel, not sequentially, so it's not waiting for an 'ok' from each one, that I can say for sure.

I'm currently using this approach in an average-size project running on Node.js, in the following scenario:

I've build a number of self-contained select components that pre-fetch data asynchronously with server actions when they get mounted. I know some people would say this is an API routes use-case, but since I'm coming from a tRPC background (being a contributor to the ecosystem and the author of [tRPC-SvelteKit](https://icflorescu.github.io/trpc-sveltekit/)), I wasn't exactly keen on building lots of REST API endpoints for that. So naturally I looked at server-actions as a boilerplate-free alternative to tRPC. 
But to my surprise and initial disappointment, I realized those requests were batched and therefore executed inefficiently during the first load. A bit of tinkering around led me to the aforementioned solution, and here's the result:
   
![Calling multiple Next.js server actions in parallel](https://github.com/user-attachments/assets/50c0e3ec-4bb2-450c-a48c-d47fdae5c7b5)

I've introduced and artificial 5 seconds delay to better illustrate the case and as you can see, the requests are indeed executed in parallel and not sequentially.
Yes, there's a slight delay (milliseconds, and it won't get multiplied with the number of actual requests) compared to calling REST API routes, but I'm perfectly OK with a few milliseconds trade-off if it spares me the effort of creating and maintaining some REST routes which make no sense outside the context of my front-end application.

Not to mention the huge advantage of being able to co-locate the loading logic with the actual components like so:

![Organizing Next.js server actions](https://github.com/user-attachments/assets/c7fe2585-3f0f-459e-b295-e139500ff4ef)

And here's how I'm actually using the server actions in components:

![Caling Next.js server actions in components](https://github.com/user-attachments/assets/ffbfad52-4b47-4627-9bcc-173a1e745f2a)

In fact, if you ask me, this would be a perfect common real-life example scenario of why we need to be able to call multiple server actions in parallel.

Projects like [tRPC.io](https://trpc.io) or [Telefunc](https://telefunc.com/) were built for solid reasons, and when I first saw Next.js server actions, I immediately thought "here's how we can do all that without the boilerplate". Being able to call them concurrently was, until now, the missing piece.

To answer your other questions, my main concern at the moment is to be able to apply this in a self-hosted Node.js environment.
There's no reason why this shouldn't work on Vercel or other environments, but I haven't tested it yet, so I'd be grateful if you or somebody else directly interested would help with that 😉

**Reply by yonatanlavy** [reply]
your help is here!
ran it in: https://github.com/yehonatanyosefi/test-parallel-actions
locally and on vercel
when on vercel, and simulated 3g network connection this is the difference:
timeRanAction: 4092
timeRanRoute: 2070
![image](https://github.com/user-attachments/assets/92913ba8-fa7c-4eb6-b8fa-26472d9f6d90)

so takes twice as much to do the roundtrip with 2 actions that take 1 second each, rather than 2 routes that take 1 second each (using 3g network throttling).
it's enough for me to never use this.
cool idea though! will definitely keep this in mind if I have more leftover functions that I have to run and haven't yet converted to trpc/api routes.

**Reply by icflorescu** [reply]
Thanks a lot for taking the time, @yehonatanyosefi!

Meanwhile I played around and added a few scenarios in the repo test app and indeed I can see differences depending on where you're running:

Here's on Vercel:

![image](https://github.com/user-attachments/assets/bf86922c-a8ff-4df8-ba09-1d2e9e450d3a)

Here's on a StackBlitz instance:

![image](https://github.com/user-attachments/assets/c64334a4-9d58-4cd4-b82b-cdcaddd6f676)

Here's on a local Node.js environment:

![image](https://github.com/user-attachments/assets/300f336f-8ffd-431b-bb98-60f4c9fadabc)

TBH, I would have expected it to perform better on Vercel.
But in my case - running on a custom VPS - there's just a few milliseconds trade-off (vs simple API routes) and I'm more than ok with that.

**Reply by omgcx1222** [reply]
> Actually, this is quite easy to do with [icflorescu/next-server-actions-parallel](https://github.com/icflorescu/next-server-actions-parallel), if you're looking for a ready-made solution. Have a look at the repo, the idea simple enough that you can simply copy-paste it in your own projects if you don't want to add it as a dependency.

This library is great, it solved my problem perfectly, it successfully changed my server actions from sequential to parallel



![sequential](https://github.com/user-attachments/assets/5f79a86d-0a47-4185-81bb-86616f408c31)
![parallel](https://github.com/user-attachments/assets/a2e21097-33b1-474e-a7df-8269d8bd6dc1)

But I saw 
```
In a nutshell: a lot faster than default Next.js server actions, not as fast as REST API routes.
Check it in your specific deployment environment and decide if it makes sense for your project.
```
on https://github.com/icflorescu/next-server-actions-parallel?tab=readme-ov-file#how-does-it-perform
In the actual code I need to change `sleep` to fetch REST API，Is there anything wrong with my implementation?
